components/places/src/bookmark_sync/record.rs

```diff
@@ -0,0 +1,244 @@
+        };
+        Ok(record)
+    }
+}
+
+impl<'de> Deserialize<'de> for BookmarkItemRecord {
+    fn deserialize<D>(d: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        // Boilerplate to translate a synced bookmark record into a typed
+        // record.
+        let raw = RawBookmarkItem::deserialize(d)?;
+        match raw.kind.as_str() {
+            "bookmark" => {
+                return Ok(BookmarkRecord {
+                    guid: raw.guid,
+                    parent_guid: raw.parent_guid,
+                    has_dupe: raw.has_dupe.unwrap_or(false),
+                    parent_name: raw.parent_name,
+                    date_added: raw.date_added,
+                    title: raw.title,
+                    url: raw.url,
+                    keyword: raw.keyword,
+                    tags: raw.tags.unwrap_or_default(),
+                }
+                .into());
+            }
+            "query" => {
+                return Ok(QueryRecord {
+                    guid: raw.guid,
+                    parent_guid: raw.parent_guid,
+                    has_dupe: raw.has_dupe.unwrap_or(false),
+                    parent_name: raw.parent_name,
+                    date_added: raw.date_added,
+                    title: raw.title,
+                    url: raw.url,
+                    folder_name: raw.folder_name,
+                }
+                .into());
+            }
+            "folder" => {
+                return Ok(FolderRecord {
+                    guid: raw.guid,
+                    parent_guid: raw.parent_guid,
+                    has_dupe: raw.has_dupe.unwrap_or(false),
+                    parent_name: raw.parent_name,
+                    date_added: raw.date_added,
+                    title: raw.title,
+                    children: raw.children.unwrap_or_default(),
+                }
+                .into());
+            }
+            "livemark" => {
+                return Ok(LivemarkRecord {
+                    guid: raw.guid,
+                    parent_guid: raw.parent_guid,
+                    has_dupe: raw.has_dupe.unwrap_or(false),
+                    parent_name: raw.parent_name,
+                    date_added: raw.date_added,
+                    title: raw.title,
+                    feed_url: raw.feed_url,
+                    site_url: raw.site_url,
+                }
+                .into());
+            }
+            "separator" => {
+                return Ok(SeparatorRecord {
+                    guid: raw.guid,
+                    parent_guid: raw.parent_guid,
+                    has_dupe: raw.has_dupe.unwrap_or(false),
+                    parent_name: raw.parent_name,
+                    date_added: raw.date_added,
+                    position: raw.position,
+                }
+                .into());
+            }
+            _ => {}
+        }
+        // TODO: We don't know how to round-trip other kinds. For now, just
```

Comment:
Is it possible to return a custom error from `deserialize`? I couldn't figure out how to unify our errors with `D::Error`.

Reply:
I think we just need `Err(de::Error::custom(format!("unknown bookmark record type '{}'", raw.kind)))` or similar. sync15's error.rs maps serde errors to ours, so I *think* that will end up as a `JsonError`, which isn't quite correct, but probably good enough.

Reply:
I pushed a fix to this.